### PR TITLE
[SYCL] Initial ABI checks implementation

### PR DESCRIPTION
**Pull request by alexbatashev**
- Add check for symbol changes on Linux
- Check sizes for classes
- Check internal layout for handler class (via AST)
- Add initial ABI Policy Guide, that describes the goals and use cases for those checks

The set of tests is incomplete and will be further extended.

Signed-off-by: Alexander Batashev <alexander.batashev@intel.com>

### CODE REVIEW

File: sycl/test/abi/symbol_size.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+int main() {
+
+  bool HasChanged = false;
+
+  using accessor_t = accessor<int, 1, access::mode::read,
+                              access::target::global_buffer, access::placeholder::true_t>;
+  CHECK_LAYOUT(accessor_t, 32)
+  CHECK_LAYOUT(buffer<int>, 40)
```
Comment (bader) [inline sycl/test/abi/symbol_size.cpp:22]: How about "compile-only" test? E.g. https://github.com/llvm/llvm-test-suite/blob/master/ABI-Testsuite/test/basic/basic.x.

Comment (alexbatashev) [reply sycl/test/abi/symbol_size.cpp]: Compile-time checks are much better, thanks for the suggestion. However, I decided to go with `static_assert` instead of macros. Does it appeal to you?

Comment (bader) [reply sycl/test/abi/symbol_size.cpp]: LGTM. Thanks.

File: sycl/test/abi/symbol_size.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
+// RUN: %clangxx -fsycl %s -o %t
+
+// Changing symbol size is a breaking change. If it happens, refer to the ABI
+// Policy Guide for further insrtuctions on breaking ABI.
```
Comment (s-kanaev) [inline sycl/test/abi/symbol_size.cpp:4]:
```suggestion
// Policy Guide for further instructions on breaking ABI.
```

File: sycl/doc/ABIPolicyGuide.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+* `test/abi/layout*` and `test/abi/symbol_size*` are a group of tests to check
+  the internal layout of some classes. The layout tests check Clang AST for
+  changes, while symbol_size check `sizeof` for objects. Changing the class
+  layout is a breaking change.
+
+## Breaking ABI
+
+Whenever you need to change the existing ABI, please, follow these steps:
+
+1. Get approval from project maintainers to make a breaking/non-breaking change.
```
Comment (pvchupin) [inline sycl/doc/ABIPolicyGuide.md:86]: We probably don't need this line. It's codeowners job to approve/don't approve.

Comment (alexbatashev) [reply sycl/doc/ABIPolicyGuide.md:86]: @pvchupin the spirit of this line is "ABI stability is a major thing, make sure reviewers understand why you do it", and not the fact, that there're some formal steps needed.

Comment (pvchupin) [reply sycl/doc/ABIPolicyGuide.md:86]: Ok. Can you rephrase accordingly then?

Comment (alexbatashev) [reply sycl/doc/ABIPolicyGuide.md]: Does it sound better now?

**Comment by alexbatashev** [comment]
> Any ideas how to reduce # of exported symbols?

@bader most of these symbols are actually builtins. Can we move those to headers?

**Comment by bader** [comment]
> > Any ideas how to reduce # of exported symbols?
> 
> @bader most of these symbols are actually builtins. Can we move those to headers?

Might be a viable option, considering almost all them are either wrappers around standard C++ functions or has trivial implementation.

File: sycl/tools/abi_check.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
+import argparse
```
Comment (kbobrovs) [inline sycl/tools/abi_check.py:1]: This module needs comments - purpose, in which cases when the check is considered passed, in which - failed

File: sycl/tools/abi_check.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+
+def compare_results(ref_records, records):
+  missing_records = set(ref_records).difference(set(records))
+  new_records = set(records).difference(set(ref_records))
+
+  return (missing_records, new_records)
+
+
+def check_symbols(ref_path, target_path):
```
Comment (kbobrovs) [inline sycl/tools/abi_check.py:59]: would be good to have comments for long functions as well

File: sycl/tools/abi_check.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
+import argparse
+import os
```
Comment (kbobrovs) [inline sycl/tools/abi_check.py:2]: also, would be good to have tests for the checker tool itself

File: sycl/doc/ABIPolicyGuide.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
+# ABI Policy Guide
+
+## Intro
+
+Application Binary Interface is a contract between binary modules, that defines
+how structures and routines are accessed in machine code. Changing the ABI may
+break backwards compatibility with library for user-developed applications,
+resulting in need to rebuild such applications. The goal of this document is to
+provide guidelines for maintaining the current ABI and mechanisms of notifying
```
Comment (kbobrovs) [inline sycl/doc/ABIPolicyGuide.md:9]:
```suggestion
provide guidelines for maintaining the current ABI of the SYCL runtime library and mechanisms of notifying
```

File: sycl/doc/ABIPolicyGuide.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+When it is not necessary to mark symbols with `__SYCL_EXPORT`:
+* The `buffer` class:
+  - It is defined by the SYCL spec, but it is fully implemented in the headers.
+* The `ProgramManager` class:
+  - It is an implementation detail.
+  - It is not accessed from the header files that are available to users.
+
+## Automated ABI Changes Testing
+
+> The automated tests deal with the most commonly occurring problems, but they
```
Comment (kbobrovs) [inline sycl/doc/ABIPolicyGuide.md:63]:
```suggestion
The automated tests deal with the most commonly occurring problems, but they
```

Comment (kbobrovs) [reply sycl/doc/ABIPolicyGuide.md:63]: and 2 lines below

Comment (alexbatashev) [reply sycl/doc/ABIPolicyGuide.md:63]: @kbobrovs I did that on purpose to differentiate this piece from the rest of the text. Any ideas how we can improve its visibility?

Comment (kbobrovs) [reply sycl/doc/ABIPolicyGuide.md:64]: I see. I just thought this was a typo. No need to change then.

File: sycl/test/tools/abi_check.dump [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# RUN: not env LLVM_BIN_PATH=%llvm_build_bin_dir python %sycl_tools_src_dir/abi_check.py --mode check_symbols --reference %s %llvm_build_lib_dir/libsycl.so | FileCheck %s
```
Comment (kbobrovs) [inline sycl/test/tools/abi_check.dump:1]: Thanks. I think the basic set of test cases should include all the 5 possible outcomes you listed abive - new symbol (done), missing symbol, changed size, changed alignment and no changes.

Comment (alexbatashev) [reply sycl/test/tools/abi_check.dump:1]: @kbobrovs we can consider missing symbols also covered. This test compares dump (this file) against libsycl.so. The python script, however, does not check symbol size. There's rather a simple cpp file with a bunch of static_asserts for that, I'm not sure we need testing for that. And neither of ABI tests currently checks for alignment, so there's nothing to verify at the point.

Comment (kbobrovs) [reply sycl/test/tools/abi_check.dump:1]: I see. Then the only case uncovered for the python script is "no changes".

Does it make sense then to have e.g. symbol_size_neg.cpp which would contain something like
  `check_size<accessor_t, 31>();`
and marked XFAIL: *
?

Comment (alexbatashev) [reply sycl/test/tools/abi_check.dump:1]: The no changes case is a bit harder to implement. We would need a binary to test on.

As for testing check_size, i see no reason for that (unless you suspect a bug in `static_assert` or `sizeof`).

File: sycl/test/tools/abi_check.dump [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+# REQUIRES: linux
+
+# CHECK: The following symbols are missing from the new object file:
+# CHECK: a_new_symbol
+# CHECK: The following symbols are new to the object file:
+
+a_new_symbol
```
Comment (keryell) [inline sycl/test/tools/abi_check.dump:8]: Missing new return symbol here :-)

**Comment by alexbatashev** [comment]
> Interesting.
> Have you chosen a simple ABI numbering easy to look at which is bumped at each ABI change?

This PR only introduces ABI checks, without versioning scheme. The part of the documentation, that says the version must be updated, should be ignored for now.

**Comment by bader** [comment]
@alexbatashev, could you merge your branch with the tip of the sycl branch, brach, please?
We just merged https://github.com/intel/llvm/commit/937fec14aeac2607af98450ddf71252321db5573 and I'd like to make sure it doesn't conflict with this patch.
